File: p/push-dominoes.rs
impl Solution {
  pub fn push_dominoes(dominoes: String) -> String {
    let n = dominoes.len();
    let chars: Vec<char> = dominoes.chars().collect();
    let mut ans = vec!['.'; n];

    let mut li = 0;
    let mut lc = ' ';
    for i in 0..n {
      let c = chars[i];
      if c == 'L' {
        if lc == 'R' {
          let w = (i - li + 1) / 2;
          for i in li..li + w { ans[i] = 'R'; }
          for i in i - w + 1..=i { ans[i] = 'L'; }
        }
        else {
          for i in li..=i { ans[i] = 'L'; }
        }
        li = i;
        lc = c;
      }
      if c == 'R' {
        if lc == 'R' {
          for i in li..=i { ans[i] = 'R'; }
        }
        li = i;
        lc = c;
      }
    }
    if lc == 'R' {
      for i in li..n { ans[i] = 'R'; }
    }

    return ans.into_iter().collect();
  }
}

File: p/push-dominoes.cc
const int INF = 0x3f3f3f3f;

class Solution {
public:
  string pushDominoes(string dominoes) {
    int N = dominoes.size();
    vector<int> lcnt(N, INF), rcnt(N, INF);
    for (int i = N - 1, flag = 0; i >= 0; --i) {
      if (dominoes[i] == 'L') lcnt[i] = 0, flag = 1;
      else if (dominoes[i] == 'R') flag = 0;
      else if (flag) lcnt[i] = lcnt[i + 1] + 1;
    }
    for (int i = 0, flag = 0; i < N; ++i) {
      if (dominoes[i] == 'R') rcnt[i] = 0, flag = 1;
      else if (dominoes[i] == 'L') flag = 0;
      else if (flag) rcnt[i] = rcnt[i - 1] + 1;
    }

    string ans(N, '.');
    for (int i = 0; i < N; ++i) {
      if (lcnt[i] < rcnt[i]) ans[i] = 'L';
      if (lcnt[i] > rcnt[i]) ans[i] = 'R';
    }
    return ans;
  }
};
